### ISSUE TITLE
Use willTransitionTo hook in rc3 - replaceWith() error

### ISSUE THREAD

**Issue by Flo-Schield-Bobby** [rackt/react-router#2326]
Hey,

(Not sure if this issue is more related to ````react-router```` or ````history```` package...)

I've read the doc about how to use the ````willTransitionTo```` hook in react-router >= 1.0.0 :
https://github.com/rackt/react-router/blob/master/UPGRADE_GUIDE.md#willtransitionto-and-willtransitionfrom

But I've having an issue while executing the ````replaceWith```` method of the ````willTransitionTo```` hook.
Here is a scheme of my use-case :

The router :
```js
    // router.js
    import React from 'react/addons';
    import createBrowserHistory from 'history/lib/createBrowserHistory';
    import Router, { Route, IndexRoute } from 'react-router';
    import AuthenticatedApp from './components/AuthenticatedApp';
    import Home from './components/authenticated/Home';

    const routes = (
        <Route path='/' component={ AuthenticatedApp }>
            // [...]
            <Route path='/home' component={ Home } onEnter={ Home.willTransitionTo } />
        </Route>
    );

    let history = createBrowserHistory();

    React.render(<Router history={ history }>{ routes }</Router>, container);
```
The High Order Component which handle ````willTransitionTo````
```js
    // components/AuthenticatedComponent.js
    import React, { Component } from 'react';
    import UserStore from '../stores/UserStore';

    export default (ComposedComponent) => {
        return class AuthenticatedComponent extends Component {
            static willTransitionTo(transition, replaceWith) {
                if (!UserStore.isLoggedIn()) {
                    replaceWith({
                        nextPathname: transition.location.pathname
                    }, '/login');
                }
            }

            // [...]
        }
    }
```
A component which requires authentication...
```js
    // components/authenticated/home.js
    import React, { Component } from 'react';
    import AuthenticatedComponent from '../AuthenticatedComponent';

    export default AuthenticatedComponent(class Home extends Component {
        // [...]
    });
```

I currently have a problem in the ````useQueries.js```` file, line 89 (````history```` package) :

````Uncaught TypeError: Cannot read property 'apply' of undefined````
```js
    function createLocation() {
        return addQuery(history.createLocation.apply(history, arguments));
    }
```
Am I missing something?
Anyway, I think an exemple of how to correctly handle this hook would be much usefull, especialy with the arguments we can pass to this ````replaceWith```` method :)

Thanks in advance!

**Comment by taion** [comment]
The auth example demonstrates exactly the pattern you're trying to use.

**Comment by Flo-Schield-Bobby** [comment]
Well, this example is the one I use as a bootstrap...

I've also tried with a method outside of the Component :

    function requireAuth(nextState, replaceState) {
        replaceState({
            nextPathname: nextState.location.pathname
        }, '/login');
    }

    const routes = (
        <Route path='/' component={ AuthenticatedApp }>
            <Route path='/home' component={ Home } onEnter={requireAuth} />
        </Route>
    );

But it doesn't work better.

**Comment by taion** [comment]
It may very well be. I couldn't possibly comment. If you start from the example, though, you ought to be able to figure out what's going on.

**Comment by Flo-Schield-Bobby** [comment]
I think I've found it.

I was using ````react```` 0.13.3 (didn't pay attention), and with just a small update to 0.14, it seems to work now...
Sorry for the confusion.

However, would it be possible to have a list of the possible parameters to pass to this ````replaceState()```` method? Is it just a basic ````history```` ````replaceState()```` call?

**Comment by opatut** [comment]
It seems like there are 3 possible arguments: `onEnter(nextState, replaceState, callback)`.

If your function does not use the callback parameter, it is considered synchronous and the callback is called automatically after execution: https://github.com/rackt/react-router/blob/master/modules/TransitionUtils.js#L5

So you could actually create an asynchronous version of onEnter. Also note that the hook function is called with its route object as `this` context.

**Comment by Flo-Schield-Bobby** [comment]
Oh, thanks, didn't find that!

**Comment by refast** [comment]
I had the same problem. Solved by updating history package from 1.9.1 to 1.13.0.
